File: baekjoon-Algorithm/algorithm-middle/dynamic-programming/N12865.py
import sys

N, K = map(int, sys.stdin.readline().split())
items = []

for _ in range(N):
    items.append(list(map(int, sys.stdin.readline().split())))

dp = [[0] * (K + 1) for _ in range(N)]

for i in range(N):
    w, v = items[i]
    for j in range(1, K+1):
        if j >= w:
            dp[i][j] = max(dp[i-1][j], dp[i-1][j-w] + v)
        else:
            dp[i][j] = dp[i-1][j]

print(dp[N-1][K])

File: baekjoon-Algorithm/algorithm-middle/dynamic-programming/N12865.java
import java.io.*;
import java.util.*;

public class N12865 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] st = br.readLine().split(" ");
        int N = Integer.parseInt(st[0]);
        int K = Integer.parseInt(st[1]);
        int[][] items = new int[N][2];

        for (int i = 0; i < N; i++) {
            st = br.readLine().split(" ");
            items[i][0] = Integer.parseInt(st[0]);
            items[i][1] = Integer.parseInt(st[1]);
        }

        int[][] dp = new int[N+1][K+1];
        int a = 0;
        int b = 0;

        for (int i = 1; i < N+1; i++) {
            int w = items[i-1][0];
            int v = items[i-1][1];

            for (int j = 1; j < K+1; j++) {
                if (j >= w) {
                    dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j-w] + v);
                } else {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }

        System.out.println(dp[N][K]);

    }
}
